File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        let found
        const dfs = (tree, i) => {
            if (tree == null) return i

            const l = dfs(tree.left, i)
            let jl = 1 + l
            console.log({ v: tree.val, i, jl, l })
            if (jl === k) {
                found = tree.val
                return jl
            }
            const jr = dfs(tree.right, jl)
            console.log({ v: tree.val, i, jr, l })
            return jr
        }
        const j = dfs(root, 0)
        console.log(j)
        return found
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
  std::vector<std::vector<int>> levelOrder(TreeNode *root) {
    auto result = std::vector<std::vector<int>>();
    auto queue = std::queue<std::pair<int, TreeNode *>>();

    if (root != nullptr)
      queue.push(std::make_pair(0, root));
    while (!queue.empty()) {
      auto [level, node] = queue.front();

      if (result.size() <= level)
        result.push_back(std::vector<int>());
      auto line = &result[level];
      line->push_back(node->val);

      queue.pop();
      if (node->left != nullptr)
        queue.push(std::make_pair(level + 1, node->left));
      if (node->right != nullptr)
        queue.push(std::make_pair(level + 1, node->right));
    }

    return result;
  }
};
